notebooks/query-dependent link analysis.ipynb

```diff
@@ -0,0 +1,1064 @@
+    "\n",
+    "# List of node property 'name'\n",
+    "nodes_graph = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    nodes_graph.append(v['properties']['name'])\n",
+    "\n",
+    "# Which nodes are in sab_pages but not in the graph\n",
+    "[node for node in sab_pages if node not in nodes_graph]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d7a80d8f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Subgraph components: [274, 16, 3, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
+    "nx.is_weakly_connected(g)\n",
+    "[len(c) for c in sorted(nx.weakly_connected_components(g), key=len, reverse=True)]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a93854ed",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "list(nx.weakly_connected_components(g))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a090b285",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Max shortest path length: 17\n",
+    "max([max(j.values()) for (i,j) in nx.shortest_path_length(g)]) # find the largest diameter amongst all components within g"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "823e754d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Between centrality: the number of shortest paths that pass through the node - a 'bridge' between nodes\n",
+    "g_di = nx.DiGraph(g)\n",
+    "g_di_between = nx.betweenness_centrality(g_di)\n",
+    "g_between = dict(sorted(g_di_between.items(), key=operator.itemgetter(1),reverse=True))\n",
+    "\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name']) # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
+    "g_between = {v: node_names_dict.get(k, v) for k, v in g_between.items()}  # replace cid with node name \n",
+    "list(g_between.items())[:10][:10]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "db344d6a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Degree centrality: counts the number of incoming and outgoing relationships from a node - 'most connected'\n",
+    "g_degree = nx.degree_centrality(g)\n",
+    "g_degree = dict(sorted(g_degree.items(), key=operator.itemgetter(1),reverse=True))\n",
+    "\n",
+    "[node[1]['properties']['name'] for node in G.nodes(data=True)]\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name'])  # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
+    "g_degree = {v: node_names_dict.get(k, v) for k, v in g_degree.items()}  # replace cid with node name \n",
```

Comment:
As above, and again for similar bits of code like this

Reply:
Thanks again, refactored in commit a3a488951396dda849cfc7c6de72e78d07499605 and